spec/controllers/external_authors_controller_spec.rb

```diff
@@ -71,91 +71,299 @@
+          expect(response).to render_template :edit
+          expect(flash[:error]).to eq "There were problems saving your preferences."
+        end
+      end
+    end
+
+    context "when the user has permission through an invitation" do
+      context "when not modifying preferences" do
+        context "when doing nothing with imported works" do
+          it "does not mark invitation redeemed and redirects with a success message" do
+            parameters = {
+              invitation_token: invitation.token,
+              id: external_author.id,
+              imported_stories: "nothing",
+              external_author: {
+                do_not_email: external_author.do_not_email,
+                do_not_import: external_author.do_not_import
+              }
+            }
+
+            put :update, params: parameters
+            it_redirects_to_with_notice(root_path, "Okay, we'll leave things the way they are! You can use the email link any time if you change your mind. Your preferences have been saved.")
+            invitation.reload
+            expect(invitation.invitee).to be_nil
+            expect(invitation.redeemed_at).to be_nil
+          end
+        end
+
+        context "when successfully updating preferences" do
+          it "does not mark invitation redeemed, saves preferences, and redirects with a success message" do
+            parameters = {
+              invitation_token: invitation.token,
+              id: external_author.id,
+              imported_stories: "nothing",
+              external_author: {
+                do_not_email: 1,
+                do_not_import: 1
+              }
+            }
+
+            put :update, params: parameters
+            it_redirects_to_with_notice(root_path, "Okay, we'll leave things the way they are! You can use the email link any time if you change your mind. Your preferences have been saved.")
+            invitation.reload
+            expect(invitation.invitee).to be_nil
+            expect(invitation.redeemed_at).to be_nil
+            external_author.reload
+            expect(external_author.do_not_email).to be_truthy
+            expect(external_author.do_not_import).to be_truthy
+          end
+        end
+
+        context "when unsuccessfully updating preferences" do
+          it "does not mark invitation redeemed and renders edit template with a success message for doing nothing and an error for preferences" do
+            parameters = {
+              invitation_token: invitation.token,
+              id: external_author.id,
+              imported_stories: "nothing",
+              external_author: {
+                do_not_email: 1,
+                do_not_import: 1
+              }
+            }
+
+            allow_any_instance_of(ExternalAuthor).to receive(:update_attributes).and_return(false)
+            put :update, params: parameters
+            allow_any_instance_of(ExternalAuthor).to receive(:update_attributes).and_call_original
```

Comment:
You can leave out the reset calls - stubs get undone at the end of a test anyway.